File: QuantConnect.AlphaStream/Models/Orders/OrderEvent.cs
using System;
using System.ComponentModel;
using System.Text;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using QuantConnect.AlphaStream.Infrastructure;

namespace QuantConnect.AlphaStream.Models.Orders
{
    public class OrderEvent
    {
        /// <summary>
        /// The unique order event id
        /// </summary>
        [JsonProperty("id")]
        public string Id => $"{AlgorithmId}-{OrderId}-{OrderEventId}";

        /// <summary>
        /// Algorithm Id, BacktestId or DeployId
        /// </summary>
        [JsonProperty("algorithm-id")]
        public string AlgorithmId { get; set; }

        /// <summary>
        /// Id of the order this event comes from.
        /// </summary>
        [JsonProperty("order-id")]
        public int OrderId { get; set; }

        /// <summary>
        /// The unique order event id for each order
        /// </summary>
        [JsonProperty("order-event-id")]
        public int OrderEventId { get; set; }

        /// <summary>
        /// Easy access to the order symbol associated with this event.
        /// </summary>
        [JsonProperty("symbol")]
        public string Symbol { get; set; }

        /// <summary>
        /// The time of this event in unix timestamp
        /// </summary>
        [JsonProperty("time"), JsonConverter(typeof(DoubleUnixSecondsDateTimeJsonConverter))]
        public DateTime Time { get; set; }

        /// <summary>
        /// Status message of the order.
        /// </summary>
        [JsonProperty("status"), JsonConverter(typeof(StringEnumConverter), true)]
        public OrderStatus Status { get; set; }

        /// <summary>
        /// The fee amount associated with the order
        /// </summary>
        [JsonProperty("order-fee-amount")]
        public decimal OrderFeeAmount { get; set; }

        /// <summary>
        /// The fee currency associated with the order
        /// </summary>
        [JsonProperty("order-fee-currency")]
        public string OrderFeeCurrency { get; set; }

        /// <summary>
        /// Fill price information about the order
        /// </summary>
        [JsonProperty("fill-price")]
        public decimal FillPrice { get; set; }

        /// <summary>
        /// Currency for the fill price
        /// </summary>
        [JsonProperty("fill-price-currency")]
        public string FillPriceCurrency { get; set; }

        /// <summary>
        /// Number of shares of the order that was filled in this event.
        /// </summary>
        [JsonProperty("fill-quantity")]
        public decimal FillQuantity { get; set; }

        /// <summary>
        /// Order direction.
        /// </summary>
        [JsonProperty("direction"), JsonConverter(typeof(StringEnumConverter), true)]
        public OrderDirection Direction { get; set; }

        /// <summary>
        /// Any message from the exchange.
        /// </summary>
        [DefaultValue(""), JsonProperty("message", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public string Message { get; set; }

        /// <summary>
        /// True if the order event is an assignment
        /// </summary>
        [JsonProperty("is-assignment")]
        public bool IsAssignment { get; set; }

        /// <summary>
        /// The current order quantity
        /// </summary>
        [JsonProperty("quantity")]
        public decimal Quantity { get; set; }

        /// <summary>
        /// The current stop price
        /// </summary>
        [JsonProperty("stop-price", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public decimal? StopPrice { get; set; }

        /// <summary>
        /// The current limit price
        /// </summary>
        [JsonProperty("limit-price", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public decimal? LimitPrice { get; set; }

        /// <summary>
        /// Returns a string that represents the current object.
        /// Allows specifying if it should also add complete ID and Symbol. It's useful not to add them when called from parent Order
        /// </summary>
        public string ToString(bool extended)
        {
            var stringBuilder = new StringBuilder();

            if (extended)
            {
                stringBuilder.Append($"Time: {Time} ID: {Id} Symbol: {Symbol} Status: {Status} Quantity {Quantity}");
            }
            else
            {
                stringBuilder.Append($"Time: {Time} OrderId {OrderEventId} Status: {Status} Quantity {Quantity}");
            }

            if (FillQuantity != 0)
            {
                stringBuilder.Append($" FillQuantity: {FillQuantity} FillPrice: {FillPrice} {FillPriceCurrency}");
            }

            if (LimitPrice.HasValue)
            {
                stringBuilder.Append($" LimitPrice: {LimitPrice.Value}");
            }
            if (StopPrice.HasValue)
            {
                stringBuilder.Append($" StopPrice: {StopPrice.Value}");
            }

            if (OrderFeeAmount != 0m) stringBuilder.Append($" OrderFee: {OrderFeeAmount} {OrderFeeCurrency}");

            if (!string.IsNullOrEmpty(Message))
            {
                stringBuilder.Append($" Message: '{Message}'");
            }

            return stringBuilder.ToString();
        }

        /// <summary>
        /// Returns a string that represents the current object.
        /// </summary>
        public override string ToString()
        {
            return ToString(true);
        }
    }
}

File: AlphaStream/Models/OrderEvent.py
from .OrderEnums import *
from datetime import datetime


class OrderEvent:
    def __init__(self, json):

        # Unique hash of the algorithm
        self.AlgorithmId = json.get('algorithm-id')

        # Unique hash of the source of the order that the order event is associated with
        self.OrderId = json.get('order-id')

        # Unique hash of the order event
        self.OrderEventId = json.get('order-event-id')

        # Unique hash of the source of the order event
        self.Id = f"{self.AlgorithmId}-{self.OrderId}-{self.OrderEventId}"

        # QuantConnect identifier for the order asset
        self.Symbol = json.get('symbol')

        # Unix time when the order event occurred
        self.Time = datetime.utcfromtimestamp(json.get('time'))

        # Fill-status of the order - new, submitted, partially filled, filled, canceled, none, invalid, cancel pending, update submitted
        self.Status = OrderStatus(json.get('status'))

        # Total fees for the order
        self.OrderFeeAmount = json.get('order-fee-amount')

        # Currency of the order fees
        self.OrderFeeCurrency = json.get('order-fee-currency')

        # Order fill price
        self.FillPrice = json.get('fill-price')

        # Currency of the order price
        self.FillPriceCurrency = json.get('fill-price-currency')

        # Number of shares filled
        self.FillQuantity = json.get('fill-quantity')

        # Direction of the order - buy, sell, hold
        self.Direction = OrderDirection(json.get('direction'))

        # Message attached to the order conveying additional information about the order event by the author
        self.Message = json.get('message', '')

        # Boolean if the order is an option assignment
        self.IsAssignment = json.get('is-assignment')

        # Number of shares of the order
        self.Quantity = json.get('quantity')

        # Stop price set for the order
        self.StopPrice = json.get('stop-price')

        # Limit price set for the order
        self.LimitPrice = json.get('limit-price')


    def __repr__(self, extended = True):

        if (extended):
            rep = f'Time: {self.Time} ID: {self.Id} Symbol: {self.Symbol} Status: {self.Status.name} Quantity: {self.Quantity}'
        else:
            rep = f'Time: {self.Time} OrderEventId: {self.OrderEventId} Status: {self.Status.name} Quantity: {self.Quantity}'

        if self.FillQuantity != 0:
            rep += f' Fill Quantity: {self.FillQuantity} Fill Price: {self.FillPrice} {self.FillPriceCurrency}'

        if self.LimitPrice is not None:
            rep += f' Limit Price: {self.LimitPrice}'

        if self.StopPrice is not None:
            rep += f' Stop Price: {self.StopPrice}'

        if self.OrderFeeAmount != 0:
            rep += f' Order Fee: {self.OrderFeeAmount} {self.OrderFeeCurrency}'

        if len(self.Message) > 0:
            rep += f' Message: "{self.Message}"'

        return rep